### PR TITLE
Nasbench201 API

### PR DESCRIPTION
**Pull request by young917**
기존 nasbench 201에서 epoch에 따른 정확도(accuracy), 걸린시간, 손실(loss)가 저장되었었는데요,
정확도는 max 값으로 하나만 저장하고, 걸린시간은 sum해서 dataset/NAS-BENCH-201-v2.pth 로 저장했습니다.
여기에 추가로 손실값을 아예 없앤 것은 dataset/NAS-BENCH-201-v2-wo-losses.pth 로 저장했습니다.

이 데이터를 로드하는 함수가 nas_201_api/api_201.py 입니다.
기존과 완전히 다르게 짰으며 nas_201_api/model_spec_201.py 의 구조체를 사용하고 있습니다.

SemiNAS에서 필요한 함수를 nasbench201과 101 에서 통일해서 쓸 수 있도록 만들었고,
이에 대한 예시로 How to use를 만들어보았습니다!

### CODE REVIEW

**Review by juice500ml — CHANGES_REQUESTED** [review]
Please add command line tool for generating smaller `.pth`.
Example: https://github.com/Multi-Objective-NAS/self-supervised-nas/blob/master/datasets/nasbench/tfrecord_sampler.py

You'll have too add in the `self-supervised-nas` repo, so please add a commit ref in the issue comment section.

File: nas_201_api/reduce_file.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+from nas_201_api import NASBench201API as API
```
Comment (juice500ml) [inline nas_201_api/reduce_file.py]: Unused import

File: nas_201_api/reduce_file.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    parser.add_argument('-o', '--output_pth_path',
+                        required=True, help='Output pth path')
+    
+    args = parser.parse_args()
+
+    _sample_and_write_pth(
+        args.number_of_rows, args.input_pth_path, 
+        args.output_pth_path, args.need_loss)
+
+
```
Comment (juice500ml) [inline nas_201_api/reduce_file.py]: This file should be inside self-supervised-nas, like https://github.com/Multi-Objective-NAS/self-supervised-nas/tree/master/datasets/nasbench

**Review by juice500ml — CHANGES_REQUESTED** [review]

File: nas_201_api/model_spec_201.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        self.ops = copy.deepcopy(ops)
+
+        self._prune()          
+        
+        self.num_nodes = np.shape(self.matrix)[0]
+        self.num_edges = np.sum(self.matrix)
+        
+        self.data_format = data_format
+    
+    def _convert_to_matrix_ops(self, arch_str):
```
Comment (juice500ml) [inline nas_201_api/model_spec_201.py]: You should add unittests with https://docs.python.org/3/library/unittest.html for this function!

File: nas_201_api/api_201.py [hunk 4]
```diff
@@ -9,383 +9,206 @@
...
-                metric2str(valid_info["loss"], valid_info["accuracy"]),
-                metric2str(test__info["loss"], test__info["accuracy"]),
-            )
-        strings += [str1, str2]
-    if show:
-        print("\n".join(strings))
-    return strings
+class NASBench201API(object):
+    def __init__(self, path: Optional[Union[Text, Dict]] = None, verbose: bool = True):
+        self.filename = None
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: On normal occasions, you shouldn't do this. Is there a reason for this?

Comment (juice500ml) [reply nas_201_api/api_201.py]: Is there a reason for keeping `self.filename`? It doesn't seem to be used elsewhere except initialization of `path`.

File: nas_201_api/api_201.py [hunk 5]
```diff
@@ -9,383 +9,206 @@
...
+        self.search_space = [
+            "skip_connect",
+            "nor_conv_1x1",
+            "nor_conv_3x3",
+            "avg_pool_3x3",
+        ]
+        self.total_time = 0
+        self.total_epochs = 0  # not implement yet
+        self.max_nodes = 0
+        self.max_edges = 0
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Initializer and declaration should be together. This is not a C++ codebase.

File: nas_201_api/api_201.py [hunk 6]
```diff
@@ -9,383 +9,206 @@
...
+
+        for archidx in sorted(list(path["arch2infos"].keys())):
+            self.arch2infos_dict[archidx] = path["arch2infos"][archidx]
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Why is `self.arch2infos_dict` OrderedDict? I can't see any specific reason for that. Please clarify.

File: nas_201_api/api_201.py [hunk 7]
```diff
@@ -9,383 +9,206 @@
...
+        archstr = self.hash2archstr[hash]
+        arch = ModelSpec(model_str=archstr, index=self.archstr2index[archstr])
+        return arch
+
+    def get_modelspec(self, matrix, ops):
+        model_spec = ModelSpec(matrix=matrix, ops=ops)
+
+        hash_val = model_spec.hash_spec()
+        if hash_val not in self.hash2archstr.keys():
+            return False
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Is this returning `False` or `model_spec`? This is quite misleading. This should be something like
```
def is_valid(self, matrix, ops):
  # Various checking procedures, may return False
  return True

def get_modelspec(self, matrix, ops):
  # If there exists a problem, it is the user's fault. Just raise exceptions and die.
  return model_spec
```

Comment (juice500ml) [reply nas_201_api/api_201.py]: The `get_modelspec` function has two responsibilities, which shouldn't happen. Split responsibilities, and give adequate amount of responsibility to the already-existing `is_valid` function.

File: nas_201_api/api_201.py [hunk 8]
```diff
@@ -9,383 +9,206 @@
...
-                op, idx = xi.split("~")
-                if op not in search_space:
-                    raise ValueError(
-                        "this op ({:}) is not in {:}".format(op, search_space)
-                    )
-                op_idx, node_idx = search_space.index(op), int(idx)
-                matrix[i + 1, node_idx] = op_idx
-        return matrix
 
+    """
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Please do not leave code comments like this.

File: nas_201_api/api_201.py [hunk 9]
```diff
@@ -9,383 +9,206 @@
...
+            avg_test_acc = 0
+
+            for seed in seeds:
+                info = archresult[(dataset, seed)]
+                test_name = [n for n in info["eval_names"] if "test" in n][0]
+                avg_test_acc += info["eval_acc1es"][test_name]
+
+            avg_test_acc /= len(seeds)
+
+            return avg_test_acc
```
Comment (juice500ml) [inline nas_201_api/api_201.py:186]: Why don't this raises exception if `option` parameter is wrong? Kill the program if parameters are wrong. Do not go silent!

File: nas_201_api/api_utils.py [hunk 10]
```diff
@@ -973,6 +973,190 @@ def _internal_query(xname):
...
+
+    def update_train_info(
+        self, train_acc1es, train_acc5es, train_losses, train_times
+    ) -> None:
+        self.train_acc1es = train_acc1es
+        self.train_acc5es = train_acc5es
+        self.train_losses = train_losses
+        self.train_times = train_times
+
+    def reset_pseudo_train_times(self, estimated_per_epoch_time: float) -> None:
```
Comment (juice500ml) [inline nas_201_api/api_utils.py]: Why is there duplicate codes here???

File: nas_201_api/reduce_file.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        for idx, t in enumerate(type_list):
+            all_results = data['arch2infos'][xkey][t]['all_results']
+            epoch = epochs[idx]
+            
+            for key in all_results.keys():
+                dataset, seed = key
+                result = all_results[key]
+
+                if isinstance(result['train_acc1es'],dict):
+                    result['train_acc1es'] = max(result['train_acc1es'].values())
```
Comment (juice500ml) [inline nas_201_api/reduce_file.py]: What happens if this is not a `dict` object? Why is this code necessary?

**Review by juice500ml — CHANGES_REQUESTED** [review]
This PR should not be merged to `master`, as originally told. Make another branch named `self-supervised-nas`, and merge it there.

**Comment by juice500ml** [comment]
`query_option` is a misnomer, as it sounds like you are querying between options. Please consider renaming it. Plain `query()` will be nicer.

File: nas_201_api/api_201.py [hunk 12]
```diff
@@ -9,266 +9,184 @@
...
+            raise OutOfDomainError('first operation should be \'input\'')
+        if model_spec.ops[-1] != 'output':
+            raise OutOfDomainError('last operation should be \'output\'')
+        for op in model_spec.ops[1:-1]:
+            if op not in self.search_space:
+                raise OutOfDomainError('unsupported op %s (available ops = %s)'
+                                                             % (op, self.search_space))
+                
+    '''
+    # matrix, ops -> looking into all archs and find isomorphic network.
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Remove comments!

File: nas_201_api/api_201.py [hunk 13]
```diff
@@ -9,266 +9,184 @@
...
-      inputs = list(filter(lambda x: x != '', node_str.split('|')))
-      for xinput in inputs: assert len(xinput.split('~')) == 2, 'invalid input length : {:}'.format(xinput)
-      for xi in inputs:
-        op, idx = xi.split('~')
-        if op not in search_space: raise ValueError('this op ({:}) is not in {:}'.format(op, search_space))
-        op_idx, node_idx = search_space.index(op), int(idx)
-        matrix[i+1, node_idx] = op_idx
-    return matrix
-
+    def query_option(self, modelspec, option, dataset='cifar10-valid'):
```
Comment (juice500ml) [inline nas_201_api/api_201.py]: Why is there '-valid' when `option` parameter exists?

File: nas_201_api/model_spec_201.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+import numpy as np
+
+# Graphviz is optional and only required for visualization.
+try:
+  import graphviz   # pylint: disable=g-import-not-at-top
+except ImportError:
+  pass
+
+
+class ModelSpec(object):
```
Comment (juice500ml) [inline nas_201_api/model_spec_201.py]: Nearly-the-same ModelSpec exists in `nasbench-101`. Please consider removing duplicate code.

File: nas_201_api/reduce_file.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
+from nas_201_api import NASBench201API as API
+import torch
+from tqdm import tqdm
+
+import argparse
+import tensorflow as tf
+import tqdm
```
Comment (juice500ml) [inline nas_201_api/reduce_file.py]: Why is there two `tqdm`s? Upper import ought to be removed.

File: nas_201_api/reduce_file.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    torch.save(data, output_path)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='Samples part of input TFrecord')
+    
+    parser.add_argument('-n', '--number_of_rows', type=int,
+                        required=True, help='Number of archs to be saved')
+
```
Comment (juice500ml) [inline nas_201_api/reduce_file.py]: Minor edit: indentation and spacing is somewhat awkward.

**Review by juice500ml — CHANGES_REQUESTED** [review]

File: nas_201_api/reduce_file.py (hunk 1 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: 없앴습니다!

File: nas_201_api/reduce_file.py (hunk 2 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: https://github.com/Multi-Objective-NAS/self-supervised-nas/blob/nasbench201/datasets/nasbench/reduce_file.py 이렇게 옮겼습니다

File: nas_201_api/api_201.py (hunk 4 above)
Comment (young917) [reply nas_201_api/api_201.py]: 말씀해주신대로 있을 필요가 없을 것 같아서 지웠습니다.

File: nas_201_api/api_201.py (hunk 5 above)
Comment (young917) [reply nas_201_api/api_201.py]: ```self.total_time``` 과 ```self.total_epochs``` 는 api를 통해 architecture의 validation이나 test 결과를 호출할 때마다 실제로 필요한 training time과 epoch를 누적시키는 값입니다. 그래서 위와 같이 하지 않고 어떻게 해야할 지 모르겠습니다...

Comment (young917) [reply nas_201_api/api_201.py]: ```self.max_nodes``` 와 ```self.max_edges``` 는 모든 architecture 을 보면서 가장 많은 노드 개수와 간선 개수를 저장하는 것입니다.
```self.max_nodes``` 대신 ```max_nodes``` 를 0으로 초기화시켜놓고 최대값을 구한 후 ```self.max_nodes = max_nodes``` 로 하였습니다.
완전히 C++ 스타일을 벗어나 의도대로 코드 짜는 방법은 잘 모르겠습니다...

File: nas_201_api/api_201.py (hunk 6 above)
Comment (young917) [reply nas_201_api/api_201.py]: 저도 이유는 모르겠지만..원래 api 에서 사용한 자료구조를 그대로 따르기 위해 사용했습니다...!
필요가 없을 것 같은데 시간만 느려질까 싶어서 OrderedDict을 사용하지 않기로 고쳤습니다!

File: nas_201_api/reduce_file.py (hunk 2 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: [see this commit!](https://github.com/Multi-Objective-NAS/self-supervised-nas/pull/2/commits/b59235c7acbd14a48ea93164c032c3fe370ed844)

File: nas_201_api/model_spec_201.py (hunk 3 above)
Comment (young917) [reply nas_201_api/model_spec_201.py]: 해보았습니다..!
[commit](https://github.com/Multi-Objective-NAS/NAS-Bench-201/pull/1/commits/c64c32e4f45fbfc5472341e7fbaac0931463a282)

https://github.com/Multi-Objective-NAS/NAS-Bench-201/blob/f8118f2e7953ed7a1d2615762cd0f3994db84ffe/nas_201_api/model_spec.py#L207

File: nas_201_api/api_utils.py (hunk 10 above)
Comment (young917) [reply nas_201_api/api_utils.py]: 기존 api에서 사용한 파일이며 저는 ```api_utils.py```를 사용하지 않습니다...!

File: nas_201_api/api_201.py (hunk 13 above)
Comment (young917) [reply nas_201_api/api_201.py]: nasbench 201 api는 cifar10-valid, cifar10, cifar100, imagenet 4가지에 대한 결과값을 다 저장하고 있습니다. 그래서 ```modelspec``` 에 대해 ```dataset``` 을 ```option```으로 훈련시킨 결과를 반환해주기 위해서 존재합니다...!

nasbench 101은 cifar10의 결과만 저장하고 있어  ```dataset```이 필요하지 않습니다.
그래서 해당 인자를 optional로 둘 수 있게 했습니다...!

File: nas_201_api/model_spec_201.py (hunk 14 above)
Comment (young917) [reply nas_201_api/model_spec_201.py]: 넵!

File: nas_201_api/api_201.py (hunk 12 above)
Comment (young917) [reply nas_201_api/api_201.py]: 네!

**Comment by young917** [comment]
> `query_option` is a misnomer, as it sounds like you are querying between options. Please consider renaming it. Plain `query()` will be nicer.

nasbench101 에 이미 query라는 함수가 있고, 이 함수를 사용하기 때문에 그랬는데요!
그래서 nasbench101 api에서
```query``` -> ```_query```
```query_option``` -> ```query```
로 수정해보았습니다...!

File: nas_201_api/reduce_file.py (hunk 11 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: ```original pth``` 파일에서,

```result['train_acc1es'] = { 0: 0.67, 1: 0.78, ... } ``` 와 같이 'epoch' , 'epoch 결과 정확도' 인  ```dict``` 형태인 경우도 있고

```result['train_acc1es'] = 0.89``` 와 같이 저장되어있는 경우도 존재해서 넣었습니다...!

File: nas_201_api/reduce_file.py (hunk 15 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: 잘못하였네요! 네!

File: nas_201_api/reduce_file.py (hunk 16 above)
Comment (young917) [reply nas_201_api/reduce_file.py]: https://github.com/Multi-Objective-NAS/self-supervised-nas/blob/master/datasets/nasbench/tfrecord_sampler.py
와 똑같이 했다고 생각했는데요...! 줄 한 칸 띄는 것을 없앴습니다!

File: nas_201_api/api_201.py (hunk 7 above)
Comment (young917) [reply nas_201_api/api_201.py]: 네! 잘 알겠습니다!!
get_modelspec 은 valid 또는 invalid 할 수 있는 modelspec 을 반환하도록 고쳤습니다!

File: nas_201_api/api_201.py (hunk 9 above)
Comment (young917) [reply nas_201_api/api_201.py:186]: 넵! 반영해보았습니다!

File: nas_201_api/api_utils.py (hunk 10 above)
Comment (juice500ml) [reply nas_201_api/api_utils.py]: Then plz remove unused codes!

File: nas_201_api/model_spec_201.py (hunk 3 above)
Comment (juice500ml) [reply nas_201_api/model_spec_201.py]: tests should be in other file!

File: nas_201_api/api_201.py (hunk 6 above)
Comment (juice500ml) [reply nas_201_api/api_201.py]: Please be sure, i think there is some code that uses dict idx! (enumeration)

File: nas_201_api/api_201.py (hunk 7 above)
Comment (juice500ml) [reply nas_201_api/api_201.py]: I don't think this is resolved yet!

File: nas_201_api/api_utils.py (hunk 10 above)
Comment (young917) [reply nas_201_api/api_utils.py]: 알겠습니다!

File: nas_201_api/api_201.py (hunk 6 above)
Comment (young917) [reply nas_201_api/api_201.py]: 제 코드에서 볼 때는 ```self.arch2infos_dict```을 key로 접근할 때만 쓰여서 필요가 없다는 생각이 듭니다....! 기존 코드에서 OrderedDict 을 사용한 게 출력할 때의 순서를 정렬시키기 위해서로 해석되었습니다...!

File: nas_201_api/api_201.py (hunk 7 above)
Comment (young917) [reply nas_201_api/api_201.py]: ```def get_modelspec(self, matrix, ops) ``` 는 아직 정말 valid 한 지 모르는 modelspec 을 반환합니다,
```def is_valid(self, modelspec)``` 은 modelspec 이 valid함을 검증하는 method로 invalid 하면 False 를 반환합니다.
그래서 ```modelspec = get_modelspec( ... )``` 후에 ```is_valid(modelspec)``` 으로 검증한 후 사용해야하는 방식으로 하였습니다.
